File: packages/desktop-helpers/macos/Sources/CorivoCaptureHelper/Screen/ScreenHandlers.swift
//
//  ScreenHandlers.swift
//  RPC handlers for `screen.capture` and `screen.list_displays`.
//

import Foundation

enum ScreenHandlers {

    private struct CapturePayload: Decodable {
        var displayId: String?
        var outputPath: String?
        var format: String
        var quality: Int?

        enum CodingKeys: String, CodingKey {
            case displayId = "display_id"
            case outputPath = "output_path"
            case format
            case quality
        }
    }

    static func capture(_ request: RequestMessage) async throws {
        guard #available(macOS 14.0, *) else {
            throw HelperError.unsupported("screen.capture requires macOS 14.0 (SCScreenshotManager)")
        }
        let payload = try decodePayload(request.payload, as: CapturePayload.self)
        guard let format = ScreenCapture.Format(rawValue: payload.format) else {
            throw HelperError.invalidRequest("unknown format: \(payload.format)")
        }
        let req = ScreenCapture.Request(
            displayId: payload.displayId,
            outputPath: payload.outputPath,
            format: format,
            quality: payload.quality
        )
        let result = try await ScreenCapture.capture(req)
        let resultJson: [String: Any] = [
            "path": result.path,
            "width": result.width,
            "height": result.height,
            "captured_at": result.capturedAt,
            "display_id": result.displayId,
        ]
        try await EventEmitter.shared.send(
            ResponseMessage(id: request.id, result: AnyCodable(resultJson))
        )
    }

    static func listDisplays(_ request: RequestMessage) async throws {
        guard #available(macOS 14.0, *) else {
            throw HelperError.unsupported(
                "screen.list_displays requires macOS 14.0 (SCShareableContent)")
        }
        let displays = try await DisplayEnumerator.list()
        let displaysArray: [[String: Any]] = displays.map { d in
            [
                "id": d.id,
                "name": d.name,
                "width": d.width,
                "height": d.height,
                "is_main": d.isMain,
            ]
        }
        let resultJson: [String: Any] = ["displays": displaysArray]
        try await EventEmitter.shared.send(
            ResponseMessage(id: request.id, result: AnyCodable(resultJson))
        )
    }

    /// Helper: decode `RequestMessage.payload` (an `AnyCodable`) into the
    /// strongly-typed payload struct. Throws `invalidRequest` on shape
    /// mismatch.
    private static func decodePayload<T: Decodable>(
        _ payload: AnyCodable?,
        as _: T.Type
    ) throws -> T {
        guard let p = payload else {
            throw HelperError.invalidRequest("missing payload")
        }
        let data: Data
        do {
            data = try JSONCodec.encoder.encode(p)
        } catch {
            throw HelperError.invalidRequest("payload reserialize failed: \(error)")
        }
        do {
            return try JSONCodec.decoder.decode(T.self, from: data)
        } catch {
            throw HelperError.invalidRequest("payload decode failed: \(error)")
        }
    }
}

File: packages/desktop-helpers/windows/src/screen/screen_handlers.cpp
// screen_handlers.cpp

#include "screen_handlers.hpp"
#include "screen_capture.hpp"
#include "display_enumerator.hpp"
#include "../ipc/event_emitter.hpp"
#include "../util/iso8601.hpp"
#include "../util/logger.hpp"

#include <stdexcept>

namespace corivo::screen {

void handle_capture(const corivo::proto::Request& request) {
    using corivo::proto::Response;
    using corivo::proto::ResponseError;
    using corivo::proto::error_codes::INVALID_REQUEST;
    using corivo::proto::error_codes::OS_ERROR;

    auto& payload = request.payload;
    if (!payload.is_object()) {
        Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{
            INVALID_REQUEST,
            "screen.capture: payload must be an object",
            nullptr,
        };
        corivo::ipc::EventEmitter::instance().send(resp);
        return;
    }

    CaptureRequest req;
    if (auto it = payload.find("display_id");
        it != payload.end() && it->is_string()) {
        req.display_id = it->get<std::string>();
    }
    if (auto it = payload.find("output_path");
        it != payload.end() && it->is_string()) {
        req.output_path = it->get<std::string>();
    }
    if (auto it = payload.find("quality");
        it != payload.end() && it->is_number_integer()) {
        req.quality = it->get<int>();
    }
    auto fmt_it = payload.find("format");
    if (fmt_it == payload.end() || !fmt_it->is_string()) {
        Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{
            INVALID_REQUEST,
            "screen.capture: missing 'format' (jpeg|png)",
            nullptr,
        };
        corivo::ipc::EventEmitter::instance().send(resp);
        return;
    }
    const std::string fmt = fmt_it->get<std::string>();
    if (fmt == "jpeg") {
        req.format = ImageFormat::Jpeg;
    } else if (fmt == "png") {
        req.format = ImageFormat::Png;
    } else {
        Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{
            INVALID_REQUEST,
            "screen.capture: unknown format: " + fmt,
            nullptr,
        };
        corivo::ipc::EventEmitter::instance().send(resp);
        return;
    }

    try {
        CaptureResponse out = capture(req);
        nlohmann::json result = {
            {"path", out.path},
            {"width", out.width},
            {"height", out.height},
            {"captured_at", corivo::iso8601::format(out.captured_at)},
            {"display_id", out.display_id},
        };
        Response resp;
        resp.id = request.id;
        resp.ok = true;
        resp.result = result;
        corivo::ipc::EventEmitter::instance().send(resp);
    } catch (const std::exception& e) {
        Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{OS_ERROR, e.what(), nullptr};
        corivo::ipc::EventEmitter::instance().send(resp);
    }
}

void handle_list_displays(const corivo::proto::Request& request) {
    using corivo::proto::Response;
    using corivo::proto::ResponseError;
    using corivo::proto::error_codes::OS_ERROR;

    try {
        auto displays = list_displays();
        nlohmann::json arr = nlohmann::json::array();
        for (auto const& d : displays) {
            arr.push_back({
                {"id", d.id},
                {"name", d.name},
                {"width", d.width},
                {"height", d.height},
                {"is_main", d.is_main},
            });
        }
        nlohmann::json result = {{"displays", arr}};
        Response resp;
        resp.id = request.id;
        resp.ok = true;
        resp.result = result;
        corivo::ipc::EventEmitter::instance().send(resp);
    } catch (const std::exception& e) {
        Response resp;
        resp.id = request.id;
        resp.ok = false;
        resp.error = ResponseError{OS_ERROR, e.what(), nullptr};
        corivo::ipc::EventEmitter::instance().send(resp);
    }
}

} // namespace corivo::screen
